src/journal/compress.c

```diff
@@ -353,6 +450,53 @@ int decompress_startswith_lz4(const void *src, uint64_t src_size,
 #endif
 }
 
+int decompress_startswith_zstd(const void *src, uint64_t src_size,
+                               void **buffer, size_t *buffer_size,
+                               const void *prefix, size_t prefix_len,
+                               uint8_t extra) {
+#if HAVE_ZSTD
+        /* Checks whether the decompressed blob starts with the
+         * mentioned prefix. The byte extra needs to follow the
+         * prefix */
+
+        int r;
+        size_t size, ret, target_buf_size = ALIGN_8(prefix_len + ZSTD_DStreamOutSize());
+        _cleanup_(ZSTD_freeDCtxp) ZSTD_DCtx *s = ZSTD_createDStream();
+
+        assert(src);
+        assert(src_size > 0);
+        assert(buffer);
+        assert(buffer_size);
+        assert(prefix);
+        assert(*buffer_size == 0 || *buffer);
+
+        if (!s)
+                return -ENOMEM;
+
+        if (!(greedy_realloc(buffer, buffer_size, target_buf_size, 1)))
+                return -ENOMEM;
+
+        ZSTD_inBuffer ZSTD_in = {src, src_size, 0};
+        ZSTD_outBuffer ZSTD_out = {*buffer, target_buf_size, 0};
```

Comment:
we usually avoid introducing variables in the middle of code, even though newer C allow that. Please declare the variable uptop, and then initialize here with:

```c
ZSTD_in = (ZSDT_inBuffer) = { src, … };
…
```